### PR TITLE
feat: GraphQL over WebSocket protocol extensions and 'connectionInit' extension

### PR DESCRIPTION
**Pull request by PacoDu**
This PR is an implementation of the solution proposed in #268 that should solve the `connection_init` handshake issue.

### CODE REVIEW

**Comment by PacoDu** [comment]
Waiting for #270, otherwise I think this PR is ready.

**Review by mcollina — COMMENTED** [review]
Good work! Can you please add a full end-to-end test where this is fully tested?

**Comment by PacoDu** [comment]
> Good work! Can you please add a full end-to-end test where this is fully tested?

This is already included in test/gateway/subscription.js (https://github.com/mcollina/fastify-gql/pull/271/files#diff-b9cdf3fa0a5c08b4c70395f7f82a9930R388). Is there anything missing ?

**Comment by mcollina** [comment]
I think we should have a test with two websocket clients to validate the end use case.

**Comment by PacoDu** [comment]
I think this is ready now 🚀 !

File: test/gateway/connectionInit-extension.js [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+      })
+
+      client.write(JSON.stringify({
+        type: 'connection_init',
+        payload: {
+          headers: {
+            authorize: true
+          }
+        }
+      }))
```
Comment (mcollina) [inline test/gateway/connectionInit-extension.js]: Can you please add _another_ client that uses a different `authorize` header (let's use `foo` and `bar`)?

Comment (PacoDu) [reply test/gateway/connectionInit-extension.js]: The part you commented is not up to date but sure i'll add a different auth method

Comment (PacoDu) [reply test/gateway/connectionInit-extension.js]: I've added 3 authentication method: one for each client and an other for gateways. Do you want me to add a third client ?

**Comment by PacoDu** [comment]
Is there anything else that prevent the merge of this PR ? I'm currently testing this version with some of our services and the `connectionInit` extension works really well.

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**